File: backend_archive/internal/repository/postgres/issue_repository.go
package postgres

import (
	"context"
	"encoding/json"
	"errors"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/shiftech/testmgr-backend/internal/domain/apperror"
	"github.com/shiftech/testmgr-backend/internal/domain/issue"
	"github.com/shiftech/testmgr-backend/internal/repository/entitycode"
	"github.com/shiftech/testmgr-backend/internal/repository/postgres/model"
)

const issueCodePrefix = "ISS"

type IssueRepository struct {
	db *gorm.DB
}

func NewIssueRepository(db *gorm.DB) *IssueRepository {
	return &IssueRepository{db: db}
}

func (r *IssueRepository) FindAllByProject(ctx context.Context, projectID string, search string, limit int) ([]issue.Issue, error) {
	tx := r.db.WithContext(ctx).Where("project_id = ?", projectID)
	if search != "" {
		like := "%" + search + "%"
		tx = tx.Where("title ILIKE ? OR code ILIKE ?", like, like)
	}
	tx = tx.Order("created_at DESC")
	if limit > 0 {
		tx = tx.Limit(limit)
	}
	var rows []model.Issue
	if err := tx.Find(&rows).Error; err != nil {
		return nil, apperror.Internal(err)
	}
	return toDomainIssues(rows)
}

func (r *IssueRepository) FindAllByTestRun(ctx context.Context, testRunID string) ([]issue.Issue, error) {
	var rows []model.Issue
	err := r.db.WithContext(ctx).
		Joins("JOIN issue_test_results itr ON itr.issue_id = issues.id").
		Joins("JOIN test_results tr ON tr.id = itr.test_result_id").
		Where("tr.test_run_id = ?", testRunID).
		Group("issues.id").
		Order("issues.created_at DESC").
		Find(&rows).Error
	if err != nil {
		return nil, apperror.Internal(err)
	}
	return toDomainIssues(rows)
}

func (r *IssueRepository) FindAllByTestResult(ctx context.Context, testResultID string) ([]issue.Issue, error) {
	var rows []model.Issue
	err := r.db.WithContext(ctx).
		Joins("JOIN issue_test_results itr ON itr.issue_id = issues.id").
		Where("itr.test_result_id = ?", testResultID).
		Order("issues.created_at DESC").
		Find(&rows).Error
	if err != nil {
		return nil, apperror.Internal(err)
	}
	return toDomainIssues(rows)
}

func (r *IssueRepository) FindByID(ctx context.Context, id string) (*issue.Issue, error) {
	var m model.Issue
	if err := r.db.WithContext(ctx).First(&m, "id = ?", id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, apperror.NotFound("issue not found")
		}
		return nil, apperror.Internal(err)
	}
	return toDomainIssue(m)
}

func (r *IssueRepository) Create(ctx context.Context, i *issue.Issue) error {
	if i.ID == "" {
		i.ID = uuid.NewString()
	}
	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		if i.Code == "" {
			code, err := entitycode.Next(ctx, tx, i.ProjectID, issueCodePrefix)
			if err != nil {
				return apperror.Internal(err)
			}
			i.Code = code
		}
		m, err := fromDomainIssue(i)
		if err != nil {
			return apperror.Internal(err)
		}
		if err := tx.Create(&m).Error; err != nil {
			if isDuplicateKeyErr(err) {
				return apperror.Conflict("an issue with this code already exists in this project")
			}
			return apperror.Internal(err)
		}
		domainIssue, err := toDomainIssue(m)
		if err != nil {
			return apperror.Internal(err)
		}
		*i = *domainIssue
		return nil
	})
}

func (r *IssueRepository) Update(ctx context.Context, i *issue.Issue) error {
	m, err := fromDomainIssue(i)
	if err != nil {
		return apperror.Internal(err)
	}
	updateErr := r.db.WithContext(ctx).Model(&model.Issue{}).Where("id = ?", i.ID).Updates(map[string]any{
		"title":           m.Title,
		"description":     m.Description,
		"actual_result":   m.ActualResult,
		"expected_result": m.ExpectedResult,
		"priority":        m.Priority,
		"type":            m.Type,
		"module_id":       m.ModuleID,
		"github_links":    m.GithubLinks,
	}).Error
	if updateErr != nil {
		return apperror.Internal(updateErr)
	}
	return nil
}

func (r *IssueRepository) UpdateStatus(ctx context.Context, id string, status issue.Status) (*issue.Issue, error) {
	if err := r.db.WithContext(ctx).Model(&model.Issue{}).Where("id = ?", id).Update("status", string(status)).Error; err != nil {
		return nil, apperror.Internal(err)
	}
	return r.FindByID(ctx, id)
}

func (r *IssueRepository) Assign(ctx context.Context, id string, assignedTo *string) (*issue.Issue, error) {
	if err := r.db.WithContext(ctx).Model(&model.Issue{}).Where("id = ?", id).Update("assigned_to", assignedTo).Error; err != nil {
		return nil, apperror.Internal(err)
	}
	return r.FindByID(ctx, id)
}

func (r *IssueRepository) Delete(ctx context.Context, id string) error {
	if err := r.db.WithContext(ctx).Delete(&model.Issue{}, "id = ?", id).Error; err != nil {
		return apperror.Internal(err)
	}
	return nil
}

func (r *IssueRepository) LinkToTestResult(ctx context.Context, issueID, testResultID string) error {
	link := model.IssueTestResult{ID: uuid.NewString(), IssueID: issueID, TestResultID: testResultID}
	err := r.db.WithContext(ctx).
		Where("issue_id = ? AND test_result_id = ?", issueID, testResultID).
		FirstOrCreate(&link).Error
	if err != nil {
		return apperror.Internal(err)
	}
	return nil
}

func (r *IssueRepository) UnlinkFromTestResult(ctx context.Context, issueID, testResultID string) error {
	err := r.db.WithContext(ctx).
		Where("issue_id = ? AND test_result_id = ?", issueID, testResultID).
		Delete(&model.IssueTestResult{}).Error
	if err != nil {
		return apperror.Internal(err)
	}
	return nil
}

func (r *IssueRepository) ReplaceTags(ctx context.Context, issueID string, tagIDs []string) error {
	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		if err := tx.Where("issue_id = ?", issueID).Delete(&model.IssueTag{}).Error; err != nil {
			return apperror.Internal(err)
		}
		if len(tagIDs) == 0 {
			return nil
		}
		rows := make([]model.IssueTag, len(tagIDs))
		for i, tagID := range tagIDs {
			rows[i] = model.IssueTag{ID: uuid.NewString(), IssueID: issueID, TagID: tagID}
		}
		if err := tx.Create(&rows).Error; err != nil {
			return apperror.Internal(err)
		}
		return nil
	})
}

func toDomainIssues(rows []model.Issue) ([]issue.Issue, error) {
	result := make([]issue.Issue, len(rows))
	for i, row := range rows {
		d, err := toDomainIssue(row)
		if err != nil {
			return nil, apperror.Internal(err)
		}
		result[i] = *d
	}
	return result, nil
}

func toDomainIssue(m model.Issue) (*issue.Issue, error) {
	var links []issue.GithubLink
	if m.GithubLinks != "" {
		if err := json.Unmarshal([]byte(m.GithubLinks), &links); err != nil {
			return nil, err
		}
	}
	if links == nil {
		links = []issue.GithubLink{}
	}
	return &issue.Issue{
		ID:             m.ID,
		ProjectID:      m.ProjectID,
		ModuleID:       m.ModuleID,
		Code:           m.Code,
		Type:           issue.Type(m.Type),
		Title:          m.Title,
		Description:    m.Description,
		ActualResult:   m.ActualResult,
		ExpectedResult: m.ExpectedResult,
		Priority:       issue.Priority(m.Priority),
		Status:         issue.Status(m.Status),
		AssignedTo:     m.AssignedTo,
		GithubLinks:    links,
		CreatedAt:      m.CreatedAt,
		UpdatedAt:      m.UpdatedAt,
	}, nil
}

func fromDomainIssue(i *issue.Issue) (model.Issue, error) {
	links := i.GithubLinks
	if links == nil {
		links = []issue.GithubLink{}
	}
	raw, err := json.Marshal(links)
	if err != nil {
		return model.Issue{}, err
	}
	return model.Issue{
		ID:             i.ID,
		ProjectID:      i.ProjectID,
		ModuleID:       i.ModuleID,
		Code:           i.Code,
		Type:           string(i.Type),
		Title:          i.Title,
		Description:    i.Description,
		ActualResult:   i.ActualResult,
		ExpectedResult: i.ExpectedResult,
		Priority:       string(i.Priority),
		Status:         string(i.Status),
		AssignedTo:     i.AssignedTo,
		GithubLinks:    string(raw),
	}, nil
}

File: frontend/src/repositories/adapters/supabase/issueRepository.ts
import { supabase } from '../../../config/supabaseClient';
import { mapAttachmentRow, mapIssueRow, mapModuleRow, mapProfileRow, mapTagRow, mapTestRoleRow } from '../../../helpers/mappers';
import type { IssueRepository } from '../../interfaces/issueRepository';
import type { Attachment, ExternalLink, Issue, IssueStatus, IssueType, IssueWithDetails } from '../../../types/domain';

const ISSUE_DETAIL_SELECT =
  '*, assignee:users!issues_assigned_to_fkey(profile:profiles(*)), module:modules(*), target_role:test_roles(*), issue_tags(tag:tags(*)), issue_test_results(test_result:test_results(id, test_run_id, test_case_code, test_case_title, test_run:test_runs(id, code, name)))';

const ISSUE_DETAIL_SELECT_INNER_LINK =
  '*, assignee:users!issues_assigned_to_fkey(profile:profiles(*)), module:modules(*), target_role:test_roles(*), issue_tags(tag:tags(*)), issue_test_results!inner(test_result:test_results!inner(id, test_run_id, test_case_code, test_case_title, test_run:test_runs(id, code, name)))';

function mapIssueWithDetailsRow(row: any): IssueWithDetails {
  return {
    ...mapIssueRow(row),
    assignee: row.assignee?.profile ? mapProfileRow(row.assignee.profile) : null,
    module: row.module ? mapModuleRow(row.module) : null,
    targetRole: row.target_role ? mapTestRoleRow(row.target_role) : null,
    tags: (row.issue_tags ?? []).map((t: any) => mapTagRow(t.tag)),
    linkedTestResults: (row.issue_test_results ?? [])
      .filter((link: any) => link.test_result)
      .map((link: any) => ({
        id: link.test_result.id,
        testRunId: link.test_result.test_run_id,
        testCaseCode: link.test_result.test_case_code,
        testCaseTitle: link.test_result.test_case_title,
        testRun: link.test_result.test_run ?? null,
      })),
  };
}

export const issueRepositoryAdapter: IssueRepository = {
  async searchByProject(projectId: string, query: string, limit = 5): Promise<Pick<Issue, 'id' | 'code' | 'title'>[]> {
    const sanitized = query.trim().replace(/^!/, '').replace(/[,()%*]/g, '');
    if (!sanitized) return [];
    const { data, error } = await supabase
      .from('issues')
      .select('id, code, title')
      .eq('project_id', projectId)
      .or(`code.ilike.%${sanitized}%,title.ilike.%${sanitized}%`)
      .limit(limit);
    if (error) throw error;
    return data ?? [];
  },

  async findByCode(projectId: string, code: string): Promise<Issue | null> {
    const { data, error } = await supabase.from('issues').select('*').eq('project_id', projectId).eq('code', code).maybeSingle();
    if (error) throw error;
    return data ? mapIssueRow(data) : null;
  },

  async findById(id: string): Promise<IssueWithDetails | null> {
    const { data, error } = await supabase.from('issues').select(ISSUE_DETAIL_SELECT).eq('id', id).maybeSingle();
    if (error) throw error;
    return data ? mapIssueWithDetailsRow(data) : null;
  },

  async findAllByProjectPaginated(
    projectId: string,
    options: {
      search?: string;
      statuses?: IssueStatus[];
      priorities?: Issue['priority'][];
      moduleIds?: string[];
      tagIds?: string[];
      types?: IssueType[];
      testRoleIds?: string[];
      page: number;
      pageSize: number;
      sortField?: string;
      sortOrder?: 'asc' | 'desc';
    },
  ): Promise<{ data: IssueWithDetails[]; total: number }> {
    const selectStr = options.tagIds?.length
      ? '*, assignee:users!issues_assigned_to_fkey(profile:profiles(*)), module:modules(*), target_role:test_roles(*), issue_tags!inner(tag:tags(*)), issue_test_results(test_result:test_results(id, test_run_id, test_case_code, test_case_title, test_run:test_runs(id, code, name)))'
      : ISSUE_DETAIL_SELECT;

    let query = supabase.from('issues').select(selectStr, { count: 'exact' }).eq('project_id', projectId);

    if (options.search?.trim()) {
      const q = options.search.trim();
      query = query.or(`title.ilike.%${q}%,code.ilike.%${q}%`);
    }
    if (options.statuses?.length) {
      query = query.in('status', options.statuses);
    }
    if (options.priorities?.length) {
      query = query.in('priority', options.priorities);
    }
    if (options.moduleIds?.length) {
      query = query.in('module_id', options.moduleIds);
    }
    if (options.testRoleIds?.length) {
      query = query.in('target_role_id', options.testRoleIds);
    }
    if (options.tagIds?.length) {
      query = query.in('issue_tags.tag_id', options.tagIds);
    }
    if (options.types?.length) {
      query = query.in('type', options.types);
    }

    const sortColumn: Record<string, string> = { code: 'code', title: 'title', priority: 'priority', status: 'status', type: 'type', assignedTo: 'assigned_to', createdAt: 'created_at', updatedAt: 'updated_at' };
    const sortCol = sortColumn[options.sortField ?? ''];
    if (options.sortField === 'moduleName') {
      query = query.order('module(name)', { ascending: (options.sortOrder ?? 'desc') === 'asc' });
    } else {
      query = query.order(sortCol ?? 'created_at', { ascending: (options.sortOrder ?? 'desc') === 'asc' });
    }

    const from = (options.page - 1) * options.pageSize;
    query = query.range(from, from + options.pageSize - 1);

    const { data, error, count } = await query;
    if (error) throw error;
    return { data: (data ?? []).map(mapIssueWithDetailsRow), total: count ?? 0 };
  },

  async findAllByProject(
    projectId: string,
    options?: { search?: string; statuses?: IssueStatus[]; priorities?: Issue['priority'][]; moduleIds?: string[]; limit?: number },
  ): Promise<IssueWithDetails[]> {
    let query = supabase
      .from('issues')
      .select(ISSUE_DETAIL_SELECT)
      .eq('project_id', projectId);

    if (options?.search?.trim()) {
      const q = options.search.trim();
      query = query.or(`title.ilike.%${q}%,code.ilike.%${q}%`);
    }
    if (options?.statuses?.length) {
      query = query.in('status', options.statuses);
    }
    if (options?.priorities?.length) {
      query = query.in('priority', options.priorities);
    }
    if (options?.moduleIds?.length) {
      query = query.in('module_id', options.moduleIds);
    }
    if (options?.limit) {
      query = query.limit(options.limit);
    }

    const { data, error } = await query.order('created_at', { ascending: false });
    if (error) throw error;
    return (data ?? []).map(mapIssueWithDetailsRow);
  },

  async findAllByTestRun(testRunId: string): Promise<IssueWithDetails[]> {
    const { data, error } = await supabase
      .from('issues')
      .select(ISSUE_DETAIL_SELECT_INNER_LINK)
      .eq('issue_test_results.test_result.test_run_id', testRunId)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return (data ?? []).map(mapIssueWithDetailsRow);
  },

  async findAllByTestResult(testResultId: string): Promise<IssueWithDetails[]> {
    const { data, error } = await supabase
      .from('issues')
      .select(ISSUE_DETAIL_SELECT_INNER_LINK)
      .eq('issue_test_results.test_result_id', testResultId)
      .order('created_at', { ascending: false });

    if (error) throw error;
    return (data ?? []).map(mapIssueWithDetailsRow);
  },

  async create(input: {
    projectId: string;
    moduleId: string | null;
    type: IssueType;
    code?: string | null;
    title: string;
    description: string | null;
    actualResult: string | null;
    expectedResult: string | null;
    priority: Issue['priority'];
    status: IssueStatus;
    assignedTo: string | null;
    targetRoleId?: string | null;
    externalLinks: ExternalLink[];
    createdBy?: string | null;
  }): Promise<Issue> {
    const { data, error } = await supabase
      .from('issues')
      .insert({
        project_id: input.projectId,
        module_id: input.moduleId,
        type: input.type,
        code: input.code || undefined,
        title: input.title,
        description: input.description,
        actual_result: input.actualResult,
        expected_result: input.expectedResult,
        priority: input.priority,
        status: input.status,
        assigned_to: input.assignedTo,
        target_role_id: input.targetRoleId ?? null,
        external_links: input.externalLinks,
        created_by: input.createdBy ?? null,
      })
      .select('*')
      .single();

    if (error) throw error;
    return mapIssueRow(data);
  },

  async createMany(
    inputs: {
      projectId: string;
      moduleId: string | null;
      type: IssueType;
      title: string;
      description: string | null;
      actualResult: string | null;
      expectedResult: string | null;
      priority: Issue['priority'];
      status: IssueStatus;
      assignedTo: string | null;
      targetRoleId?: string | null;
      externalLinks: ExternalLink[];
      createdBy?: string | null;
    }[],
  ): Promise<Issue[]> {
    if (inputs.length === 0) return [];
    const { data, error } = await supabase
      .from('issues')
      .insert(inputs.map((i) => ({
        project_id: i.projectId,
        module_id: i.moduleId,
        type: i.type,
        title: i.title,
        description: i.description,
        actual_result: i.actualResult,
        expected_result: i.expectedResult,
        priority: i.priority,
        status: i.status,
        assigned_to: i.assignedTo,
        target_role_id: i.targetRoleId ?? null,
        external_links: i.externalLinks,
        created_by: i.createdBy ?? null,
      })))
      .select('*');
    if (error) throw error;
    return (data ?? []).map(mapIssueRow);
  },

  async update(
    id: string,
    changes: Partial<
      Pick<Issue, 'code' | 'title' | 'description' | 'actualResult' | 'expectedResult' | 'priority' | 'type' | 'moduleId' | 'targetRoleId' | 'externalLinks'>
    >,
  ): Promise<Issue> {
    const payload: Record<string, unknown> = {};
    if (changes.code !== undefined) payload.code = changes.code;
    if (changes.title !== undefined) payload.title = changes.title;
    if (changes.description !== undefined) payload.description = changes.description;
    if (changes.actualResult !== undefined) payload.actual_result = changes.actualResult;
    if (changes.expectedResult !== undefined) payload.expected_result = changes.expectedResult;
    if (changes.priority !== undefined) payload.priority = changes.priority;
    if (changes.type !== undefined) payload.type = changes.type;
    if (changes.moduleId !== undefined) payload.module_id = changes.moduleId;
    if (changes.targetRoleId !== undefined) payload.target_role_id = changes.targetRoleId;
    if (changes.externalLinks !== undefined) payload.external_links = changes.externalLinks;

    const { data, error } = await supabase.from('issues').update(payload).eq('id', id).select('*').single();
    if (error) throw error;
    return mapIssueRow(data);
  },

  async updateStatus(id: string, status: IssueStatus): Promise<Issue> {
    const { data, error } = await supabase.from('issues').update({ status }).eq('id', id).select('*').single();
    if (error) throw error;
    return mapIssueRow(data);
  },

  async assign(id: string, assignedTo: string | null): Promise<Issue> {
    const { data, error } = await supabase.from('issues').update({ assigned_to: assignedTo }).eq('id', id).select('*').single();
    if (error) throw error;
    return mapIssueRow(data);
  },

  async remove(id: string): Promise<void> {
    const { error } = await supabase.from('issues').delete().eq('id', id);
    if (error) throw error;
  },

  async linkToTestResult(issueId: string, testResultId: string): Promise<void> {
    const { error } = await supabase
      .from('issue_test_results')
      .upsert({ issue_id: issueId, test_result_id: testResultId }, { onConflict: 'issue_id,test_result_id', ignoreDuplicates: true });
    if (error) throw error;
  },

  async unlinkFromTestResult(issueId: string, testResultId: string): Promise<void> {
    const { error } = await supabase
      .from('issue_test_results')
      .delete()
      .eq('issue_id', issueId)
      .eq('test_result_id', testResultId);
    if (error) throw error;
  },

  async replaceTags(issueId: string, tagIds: string[]): Promise<void> {
    const { error: deleteError } = await supabase.from('issue_tags').delete().eq('issue_id', issueId);
    if (deleteError) throw deleteError;
    if (tagIds.length === 0) return;
    const { error: insertError } = await supabase.from('issue_tags').insert(tagIds.map((tagId) => ({ issue_id: issueId, tag_id: tagId })));
    if (insertError) throw insertError;
  },

  async findAttachments(issueId: string): Promise<Attachment[]> {
    const { data, error } = await supabase
      .from('entity_attachments')
      .select('*')
      .eq('entity_type', 'issue')
      .eq('entity_id', issueId)
      .order('created_at', { ascending: true });
    if (error) throw error;
    return (data ?? []).map(mapAttachmentRow);
  },

  async addAttachment(input: {
    issueId: string;
    projectId: string;
    storageProvider: string;
    url: string;
    fileName: string;
    fileSize: number | null;
    contentType: string | null;
  }): Promise<Attachment> {
    const { data, error } = await supabase
      .from('entity_attachments')
      .insert({
        entity_type: 'issue',
        entity_id: input.issueId,
        project_id: input.projectId,
        storage_provider: input.storageProvider,
        url: input.url,
        file_name: input.fileName,
        file_size: input.fileSize,
        content_type: input.contentType,
      })
      .select('*')
      .single();
    if (error) throw error;
    return mapAttachmentRow(data);
  },

  async removeAttachment(id: string): Promise<void> {
    const { error } = await supabase.from('entity_attachments').delete().eq('id', id);
    if (error) throw error;
  },
};
